### PR TITLE
feat(api): define & execute v3 json protocols

### LINKED ISSUE #3110: Write new JSON schema for V3
This ticket should be updated to include any changes we want to make in the V3 protocol json schema. The result of this ticket will be a `protocolSchemaV3.json` file added to `shared-data/`.

This ticket does NOT include making PD save V3 protocols, or loading V3 protocols in Run App / JSON executor/ etc. This ticket is a prereq of all that work, and is just concerned with **writing the schema definition** which we will move towards.

# Changes for the next JSON schema:

* Convert all kebab-cased keys to camelCase
* Remove `pipette-model` and all pipette "model" related fields. Require "name" field for pipette, no longer optional.
* Set the version number as part of the schema (eg `schemaVersion: 3` is required, value must be the integer `3` for it to be a valid schema v3 protocol). From v3 onward, `schema-version / schemaVersion` will be an `integer`, not a semver string.
* make `touchTipMmFromTop` required under `defaultValues` - no longer optional (WIP: any missing defaultValues?)
* Make `labware` key contain labware definitions, remove `labware.model` and use new`labware.definition`
* ! Make sure all params are sealed with `"additionalProperties": false` -- unless it's impossible w/ ref reuse :/

# STUFF WE HAVEN'T ALL TALKED ABOUT: Breaking `procedure` up

Make `procedure` a flat array (call it `commands`), and use an optional `commandAnnotations` object to keep an arbitrarily-nested array of description objects that references indicies in `commands`. The `commandAnnotations` could be used for the RA run log when it is present.

Since we aren't sure what kind of data will go in there, let's leave `commandAnnotations` permissive - it's an optional object of arbitrary shape. That means to use it, the consumer (Run App) will need to be defensive, or we'll have to release a new JSON protocol schema version.

The 2 motivations are that procedures/subProcedures does not allow arbitrarily deep nesting, only one level -- but annotations in the Run Log might be better deep. And, keeping commands flat and un-annotated will make the JSON executor's task simpler.

### PR DESCRIPTION
**Pull request by IanLondon**
## overview

Closes #3110 and implements execution in APIv2

Does NOT implement execution in APIv1, I think we don't need to support that because by the time v3 JSON Protocols are out in the wild, we'll have moved to APIv2? If not, we can update the APIv1 executor later (I think...)

## changelog

* define v3 json protocol schema
* support v3 JSON protocol execution in APIv2 executor
* split JSON protocol version specific execution fns into their own .py files
* add unit tests for v3 execution fns

## review requests

NOTE: the example protocols and the schema are different than #3307. Following a suggestion from @sfoster1, `defaultValues` key has been deleted in favor of explicit per-command params

[v3jsonProtocolExamples.zip](https://github.com/Opentrons/opentrons/files/3049898/v3jsonProtocolExamples.zip)

- [ ] Review V3 schema def
- [ ] Are there other parts of the commands still using implicit values that should be make into params?
- [ ] Check out the schema, make sure it's designed in a sensible way that accomplishes the things we've talked about and doesn't block anything we need

- [ ] Play with the example v3 protocol files and make sure the schema validates them when they should and fails to validate "bad" protocol files

`npm install -g ajv-cli` then you can use the `ajv` CLI to validate v3 JSON protocols with the new schema:

`ajv validate -s shared-data/protocol-json-schema/protocolSchemaV3.json -r shared-data/labware-json-schema/labware-schema.json -d path/to/exampleProtocol.json` (the `-r` pulls in the v2 labware schema so it resolve the reference to `opentronsLabwareSchemaV2` in the definitions section)

- [ ] The V3 protocols should run correctly on the robot (with APIv2 feature flag enabled) -- we need to carefully test the offset and flow rates and make sure they're going in as intended
- [ ] Pre-V3 protocols should still run on the robot, the same as before (eg V1 protocols exported from PD)
- [ ] Custom labware created by `labware-designer` should work when embedded in these protocols!

### CODE REVIEW

File: api/src/opentrons/protocol_api/execute.py [hunk 1]
```diff
@@ -111,227 +108,25 @@ def _run_python(proto: Any, context: ProtocolContext):
...
+    # do not permit protocols with nonexistent schema versions to load
+    if (legacyKebabVersion == '1.0.0'):
+        return 1
+    if (legacyKebabVersion == '2.0.0'):
+        return 2
+    if (legacyKebabVersion is not None):
+        raise RuntimeError(('No such schema version: "{}". Did you mean ' +
+                           '"1.0.0" or "2.0.0"?').format(legacyKebabVersion))
+    raise RuntimeError(
+        'Could not determine schema version for protcol. ' +
```
Comment (sfoster1) [inline api/src/opentrons/protocol_api/execute.py]:
```suggestion
        'Could not determine schema version for protocol. ' +
```

File: api/src/opentrons/protocol_api/execute.py [hunk 2]
```diff
@@ -364,8 +159,22 @@ def run_protocol(protocol_code: Any = None,
     if None is not protocol_code:
         _run_python(protocol_code, true_context)
     elif None is not protocol_json:
-        lw = load_labware_from_json(true_context, protocol_json)
-        ins = load_pipettes_from_json(true_context, protocol_json)
-        dispatch_json(true_context, protocol_json, ins, lw)
+        protocol_version = get_protocol_schema_version(protocol_json)
+        if (protocol_version) > 3:
+            raise RuntimeError('JSON Protocol version {} is not yet supported \
```
Comment (sfoster1) [inline api/src/opentrons/protocol_api/execute.py]: May fail lint; \ is not required inside braces, you can do

```
raise RuntimError('Json Protocol version {} is not yet supported'
                              'in this version of the API')
```

File: api/src/opentrons/protocol_api/execute_v1.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+MODULE_LOG = logging.getLogger(__name__)
+
+
+def load_pipettes_from_json(
+        ctx: ProtocolContext,
+        protocol: Dict[Any, Any]) -> Dict[str, InstrumentContext]:
+    pipette_data = protocol.get('pipettes', {})
+    pipettes_by_id = {}
+    for pipette_id, props in pipette_data.items():
+        model = props.get('model')
```
Comment (sfoster1) [inline api/src/opentrons/protocol_api/execute_v1.py]: If these keys are mandatory we should access them via dict indexing: `model = props['model']` which will fail fast with a `KeyError` if the protocol is malformed and the keys aren't present, and fixes issues with types since `get()` returns `Optional[T]`

File: api/src/opentrons/protocol_api/execute_v3.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+from . import labware
+from opentrons.types import Point, Location
+
+MODULE_LOG = logging.getLogger(__name__)
+
+
+def load_pipettes_from_json(
+        ctx: ProtocolContext,
+        protocol: Dict[Any, Any]) -> Dict[str, InstrumentContext]:
+    pipette_data = protocol.get('pipettes', {})
```
Comment (sfoster1) [inline api/src/opentrons/protocol_api/execute_v3.py]: Accesses to mandatory keys should be dict indexing (`protocol['pipettes']`) to avoid typing issues and fail fast

File: api/src/opentrons/protocol_api/execute_v3.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        mount = props.get('mount')
+        name = props.get('name')
+        instr = ctx.load_instrument(name, mount)
+        pipettes_by_id[pipette_id] = instr
+
+    return pipettes_by_id
+
+
+def _get_well(loaded_labware: Dict[str, labware.Labware],
+              params: Dict[str, Any]):
```
Comment (sfoster1) [inline api/src/opentrons/protocol_api/execute_v3.py]: return type docs?

File: api/src/opentrons/protocol_api/execute_v3.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+# absolute value (not % value) as an argument in
+# aspirate/dispense/blowout/air_gap fns, remove this
+def _set_flow_rate(
+        pipette_name, pipette, command_type, params):
+    """
+    Set flow rate in uL/mm, to value obtained from command's params.
+    """
+    flow_rate_param = params.get('flowRate')
+
+    pipette.flow_rate = {
```
Comment (sfoster1) [inline api/src/opentrons/protocol_api/execute_v3.py]: Same mandatory key thing but also if it's not mandatory, then we should check the value of `flow_rate_param` so we don't assign `None` here

File: api/src/opentrons/protocol_api/execute_v3.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+        'aspirate': flow_rate_param,
+        'dispense': flow_rate_param
+    }
+
+
+def load_labware_from_json_defs(
+        ctx: ProtocolContext,
+        protocol: Dict[Any, Any]) -> Dict[str, labware.Labware]:
+    protocol_labware = protocol.get('labware', {})
+    definitions = protocol.get('labwareDefinitions', {})
```
Comment (sfoster1) [inline api/src/opentrons/protocol_api/execute_v3.py]: Mandatory keys?

File: api/src/opentrons/protocol_api/execute_v3.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+        protocol_pipette_data = protocol_data\
+            .get('pipettes', {})\
+            .get(params.get('pipette'), {})
+        pipette_name = protocol_pipette_data.get('name')
+
+        if (not pipette_name):
+            # TODO: Ian 2018-11-06 remove this fallback to 'model' when
+            # backwards-compatability for JSON protocols with versioned
+            # pipettes is dropped (next JSON protocol schema major bump)
+            pipette_name = protocol_pipette_data.get('model')
```
Comment (sfoster1) [inline api/src/opentrons/protocol_api/execute_v3.py]: This TODO is in fact resolved in v3, right?

File: api/tests/opentrons/protocol_api/test_execute_v3.py [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
...A4":{"shape":"rectangular","depth":42.16,"length":8.33,"width":71.88,"totalLiquidVolume":22000,"x":41.21,"y":42.9,"z":2.29},"A5":{"shape":"rectangular","depth":42.16,"length":8.33,"width":71.88,"totalLiquidVolume":22000,"x":50.3,"y":42.9,"z":2.29},"A6":{"shape":"rectangular","depth":42.16,"length":8.33,"width":71.88,"totalLiquidVolume":22000,"x":59.39,"y":42.9,"z":2.29},"A7":{"shape":"rectangular","depth":42.16,"length":8.33,"width":71.88,"totalLiquidVolume":22000,"x":68.48,"y":42.9,"z":2.29},"A8":{"shape":"rectangular","depth":42.16,"length":8.33,"width":71.88,"totalLiquidVolume":22000,"x":77.57,"y":42.9,"z":2.29},"A9":{"shape":"rectangular","depth":42.16,"length":8.33,"width":71.88,"totalLiquidVolume":22000,"x":86.66,"y":42.9,"z":2.29},"A10":{"shape":"rectangular","depth":42.16,"length":8.33,"width":71.88,"totalLiquidVolume":22000,"x":95.75,"y":42.9,"z":2.29},"A11":{"shape":"rectangular","depth":42.16,"length":8.33,"width":71.88,"totalLiquidVolume":22000,"x":104.84,"y":42.9,"z":2.29},"A12":{"shape":"rectangular","depth":42.16,"length":8.33,"width":71.88,"totalLiquidVolume":22000,"x":113.93,"y":42.9,"z":2.29}},"brand":{"brand":"USA Scientific","brandId":["1061-8150"]}}""")  # noqa
```
Comment (sfoster1) [inline api/tests/opentrons/protocol_api/test_execute_v3.py]: Maybe put this in a data file in this dir?

File: api/tests/opentrons/protocol_api/test_execute_v3.py [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+    def __setattr__(self, name, value):
+        if name == 'log':
+            super(MockPipette, self).__setattr__(name, value)
+        else:
+            self.log.append(("set: {}".format(name), value))
+
+
+def test_dispatch_commands(monkeypatch, loop):
+    protocol_data = {
+        "schemaVersion": "3",
```
Comment (sfoster1) [inline api/tests/opentrons/protocol_api/test_execute_v3.py]: Put it in a data file in this dir?

File: api/src/opentrons/protocol_api/execute_v3.py [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+
+
+def dispatch_json(context: ProtocolContext,  # noqa(C901)
+                     protocol_data: Dict[Any, Any],
+                     instruments: Dict[str, InstrumentContext],
+                     loaded_labware: Dict[str, labware.Labware]):
+    commands = protocol_data.get('commands', [])
+
+    for command_item in commands:
+        command_type = command_item.get('command')
```
Comment (sfoster1) [inline api/src/opentrons/protocol_api/execute_v3.py]: We could take the opportunity here to make a `CommandTypes` [enum](https://docs.python.org/3.6/library/enum.html)

File: api/src/opentrons/protocol_api/execute_v3.py [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+            .get(params.get('pipette'), {})
+        pipette_name = protocol_pipette_data.get('name')
+
+        if (not pipette_name):
+            # TODO: Ian 2018-11-06 remove this fallback to 'model' when
+            # backwards-compatability for JSON protocols with versioned
+            # pipettes is dropped (next JSON protocol schema major bump)
+            pipette_name = protocol_pipette_data.get('model')
+
+        if command_type == 'delay':
```
Comment (sfoster1) [inline api/src/opentrons/protocol_api/execute_v3.py]: It may be scope creep here but one thing the `default-params` removal makes a bit easier is the ability to refactor individual command executors into separate functions that you load into a dict keyed by command names (or command enums) and passed a uniform set of parameters. We should consider that here.

File: api/src/opentrons/protocol_api/execute.py (hunk 2 above)
Comment (IanLondon) [reply api/src/opentrons/protocol_api/execute.py]: `make -C api lint` is working for me as-is but would you prefer this?

```python
raise RuntimeError((
                'JSON Protocol version {} is not yet ' +
                'supported in this version of the API')
                .format(protocol_version))
```

File: api/src/opentrons/protocol_api/execute.py [hunk 13]
```diff
@@ -111,227 +108,25 @@ def _run_python(proto: Any, context: ProtocolContext):
...
-            pipette.move_to(  # type: ignore
-                slot_obj.move(offsetPoint),
-                force_direct=params.get('force-direct'),
-                minimum_z_height=params.get('minimum-z-height'))
-        else:
-            MODULE_LOG.warning("Bad command type {}".format(command_type))
+def get_protocol_schema_version(protocol_json: Dict[Any, Any]) -> int:
+    # v3 and above uses `schemaVersion: integer`
+    version = protocol_json.get('schemaVersion')
+    if None is not version:
```
Comment (Laura-Danielle) [inline api/src/opentrons/protocol_api/execute.py]: Could you change this to either:

```
if protocol_json.get('schemaVersion'):
    return protocol_json.get('schemaVersion')
```

or:

```
version = protocol_json.get('schemaVersion')
if version:
    return version
```

File: api/src/opentrons/protocol_api/execute.py [hunk 14]
```diff
@@ -111,227 +108,25 @@ def _run_python(proto: Any, context: ProtocolContext):
...
+        return version
+    # v1 uses 1.x.x and v2 uses 2.x.x
+    legacyKebabVersion = protocol_json.get('protocol-schema')
+    # No minor/patch schemas ever were released,
+    # do not permit protocols with nonexistent schema versions to load
+    if (legacyKebabVersion == '1.0.0'):
+        return 1
+    if (legacyKebabVersion == '2.0.0'):
+        return 2
+    if (legacyKebabVersion is not None):
```
Comment (Laura-Danielle) [inline api/src/opentrons/protocol_api/execute.py]: Same here:

```
if legacyKebabVersion:
```

Also may want to change format to if/elif as that's normally used for conditional checking

File: api/src/opentrons/protocol_api/execute.py [hunk 15]
```diff
@@ -364,8 +159,22 @@ def run_protocol(protocol_code: Any = None,
     if None is not protocol_code:
```
Comment (Laura-Danielle) [inline api/src/opentrons/protocol_api/execute.py]: Same here for checking whether a variable is populated (`if protocol_code`)

File: api/src/opentrons/protocol_api/execute.py (hunk 2 above)
Comment (Laura-Danielle) [reply api/src/opentrons/protocol_api/execute.py]: You can also do this fancy new thing (similar to javascript evaluated strings)

`f'Json Protocol version {protocol_version} is not yet supported in this version of the API'` if you don't want to use format

File: api/src/opentrons/protocol_api/execute_v1.py [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+    for command_item in flat_subs:
+        command_type = command_item.get('command')
+        params = command_item.get('params', {})
+        pipette = instruments.get(params.get('pipette'))
+        protocol_pipette_data = protocol_data\
+            .get('pipettes', {})\
+            .get(params.get('pipette'), {})
+        pipette_name = protocol_pipette_data.get('name')
+
+        if (not pipette_name):
```
Comment (Laura-Danielle) [inline api/src/opentrons/protocol_api/execute_v1.py]: don't need parentheses here

File: shared-data/protocol-json-schema/protocolSchemaV3.json [hunk 17]
```diff
@@ -0,0 +1,383 @@
...
+        }
+      }
+    },
+
+    "designerApplication": {
+      "description": "Optional data & metadata not required to execute the protocol, used by the application that created this protocol",
+      "type": "object",
+      "properties": {
+        "name": {
+          "description": "Name of the application that created the protocol. Should be namespaced under the organization or individual who owns the organization, eg \"opentrons/protocol-designer\"",
```
Comment (Laura-Danielle) [inline shared-data/protocol-json-schema/protocolSchemaV3.json]: Is the name always guaranteed to be `opentrons/protocol-designer`? Should we restrict it to that value only?

Comment (IanLondon) [reply shared-data/protocol-json-schema/protocolSchemaV3.json]: nope, it's open for any 3rd party JSON protocol creator app to use

File: api/src/opentrons/protocol_api/execute_v3.py (hunk 11 above)
Comment (sfoster1) [reply api/src/opentrons/protocol_api/execute_v3.py]: Not necessary IMO with the new map approach

File: api/src/opentrons/protocol_api/execute.py (hunk 14 above)
Comment (sfoster1) [reply api/src/opentrons/protocol_api/execute.py]: pylint yells at you if you do
```
if:
  return
else:
  other stuff
```

I personally disagree but if that's what the linter says that's what the linter says

File: api/src/opentrons/protocol_api/execute_v3.py [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+    if not (flow_rate_param > 0):
+        raise RuntimeError('Positive flowRate param required')
+
+    pipette.flow_rate = {
+        'aspirate': flow_rate_param,
+        'dispense': flow_rate_param
+    }
+
+
+def load_labware_from_json_defs(
```
[the remark this answers is not in the document]
Comment (sfoster1) [reply api/src/opentrons/protocol_api/execute_v3.py]: Yeah I'm OK with that.
